### PR TITLE
Issue #922 Ability to Disable Profile Validation (only codecov/project failing)

### PR DESCRIPTION
**Pull request by JillieBeanSim**
## Proposed changes

As a user, I would like to be able to disable the validation of profiles setting since I may be working with cloud instances (zD&Ts) or particularly slow LPARS that make this process long and unnecessary.
With that, we introduced a persistent global setting called `Zowe-Automatic-Validation` whose default is `true` and will automatically validate each profile. (Setting this to `false` will disable the validation of all profiles.) Additionally, a right-click option was added for each profile that will allow the user to override the global setting and Enable or Disable validation for that particular profile during the user's VS Code session.
See GIFs below for this issue.
https://github.com/zowe/vscode-extension-for-zowe/issues/922

Validating Profiles blocks users from accessing other profiles until the validation is done, which can sometimes take 30 secs or more. The user is not able to add another profile or run a query on another tree element while a validation is running.
See GIF below for this issue.
https://github.com/zowe/vscode-extension-for-zowe/issues/953

## Release Notes

Milestone: 1.9.0

https://github.com/zowe/vscode-extension-for-zowe/issues/922
Changelog: Introduction of a persistent global setting called `Zowe-Automatic-Validation` whose default is `true` and will automatically validate each profile. (Setting this to `false` will disable the validation of all profiles.) Additionally, a right-click option for each profile that will allow the user to override the global setting and Enable or Disable validation for that particular profile during the user's VS Code session.

https://github.com/zowe/vscode-extension-for-zowe/issues/953
Changelog(Bugfix): While validating a Profile the user will no longer be blocked from accessing other profiles.

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [x] PR Description is included
- [x] gif or screenshot is included if visual changes are made
- [x] `npm run vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [x] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [x] There is coverage for the code that I have added
- [x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments
https://github.com/zowe/vscode-extension-for-zowe/issues/922
Validation-onChange-settings
Validation-right-click

https://github.com/zowe/vscode-extension-for-zowe/issues/953
Validation-not-blocking

If your LPAR is too quick to test where the validation isn't blocking other profiles this can be added to Profiles.ts line 1032, replace that line with the snippet below to slow down the validation:
```
return new Promise((resolve) => {
     setTimeout(async () => {
         resolve(await getSessStatus.getStatus(theProfile, theProfile.type));
      // tslint:disable-next-line:no-magic-numbers
      }, 5000);
});
```

### CODE REVIEW

**Comment by jellypuno** [comment]
@JillieBeanSim Thank you for adding this. I was able to test it and found some issues

1. If I change the Global Validation to **false** and search for my datasets using the magnifying glass, the search doesn't work.
2. If I change the Global Validation to **true** and search for my datasets again, The icon is not changed.
3. Invalid Profiles changes icons but Valid Profiles doesn't change.
4. If I add another profile in the tree, the validation for all profiles added in the tree restarts and the icon is set to its default icon.

Validation Err1
Validation Err2

**Comment by jellypuno** [comment]
@JillieBeanSim 

These are the only parts that I could see that were not covered by your changes. Some are in extension.ts that we cannot actually cover. 

Uss/USSTree 
 
image


job/ZosJobsProvider
image
image

 
Dataset/DatasetTree
image
image

**Comment by jellypuno** [comment]
@JillieBeanSim Is this ready for review?

**Comment by JillieBeanSim** [comment]
> @JillieBeanSim Is this ready for review?

@jellypuno yes, thanks

**Comment by jellypuno** [comment]
Issues found:

1. Global Validation is set to OFF
- Filter dataset doesn't work
- Editing Profile calls validation
- Adding a new profile resets the validation status

validationOFF

2. Global Validation is set to ON
- I disabled the validation in a specific profile, the filtering doesn't work.

validationON

**Review by lauren-li — DISMISSED** [review]
Overall, this looks good so far! The ability to interact with other profiles while a slow profile is still validating is really awesome.

I've made some small suggestions in `package.json`/`package.nls.json` regarding wording/localization. Additionally, in `package.json`, I was wondering if we could move the “Disable/Enable validation” commands to the `“0_mainframeInteraction”` command group? I think it makes more sense for them to be placed there, as they do not seem to really "modify" the profile in the same way that other commands in their current group (`"modification"`, e.g. "Delete Profile") do.

I am still seeing the following issues @jellypuno mentioned:
- Adding a profile into the tree resets the validation status. (Actually, it appears to reset/refresh the tree as a whole. The tree's nodes all collapse, which does not seem to happen with the current Marketplace version.)
- Clicking the Saved Search in Favorites resets the validation status.

Aside from the above two, the other issues Jelly mentioned appear to be fixed!




Thanks @JillieBeanSim for your hard work on this!

**Review by crawr — COMMENTED** [review]
Testing in Theia, I also found that adding something to favorites also resets the validation.

File: i18n/sample/package.i18n.json [hunk 1]
```diff
@@ -80,5 +81,11 @@
...
     "profile.configuration.url": "z/OSMF URL",
     "profile.configuration.user": "z/OSMF User Name",
-    "profile.configuration.ru": "Profile Reject Unauthorize"
+    "profile.configuration.ru": "Profile Reject Unauthorize",
+    "disableValidation": "Disable Validation",
+    "uss.disableValidation": "Disable Validation",
+    "jobs.disableValidation": "Disable Validation",
+    "enableValidation": "Enable Validation",
+    "uss.enableValidation": "Enable Validation",
+    "jobs.enableValidation": "Enable Validation"
```
[the remark this answers is not in the document]
Comment (JillieBeanSim) [reply i18n/sample/package.i18n.json]: addressed in latest commit

File: i18n/sample/package.i18n.json [hunk 2]
```diff
@@ -80,5 +81,11 @@
     "profile.configuration.name": "Profile Name",
     "profile.configuration.url": "z/OSMF URL",
     "profile.configuration.user": "z/OSMF User Name",
-    "profile.configuration.ru": "Profile Reject Unauthorize"
+    "profile.configuration.ru": "Profile Reject Unauthorize",
+    "disableValidation": "Disable Validation",
+    "uss.disableValidation": "Disable Validation",
+    "jobs.disableValidation": "Disable Validation",
```
[the remark this answers is not in the document]
Comment (JillieBeanSim) [reply i18n/sample/package.i18n.json]: addressed in latest commit

File: package.nls.json [hunk 3]
```diff
@@ -68,6 +68,7 @@
     "Zowe-Default-Jobs-Persistent-Favorites": "Toggle if Jobs favorite files persist locally",
     "Zowe-Default-Commands-Persistent-Favorites": "Toggle if Commands persist locally",
     "Zowe-Environment": "Environment where the extension is running, default is VSCode",
+    "Zowe-Automatic-Profile-Validation": "Allow validation of profiles.",
```
[the remark this answers is not in the document]
Comment (JillieBeanSim) [reply package.nls.json]: addressed in latest commit

**Comment by JillieBeanSim** [comment]
@jellypuno @lauren-li @crawr could you all try the PR out again please?

I am having a hard time getting codeCov/patch up due to not being able to get into this if statement in data set, uss, and job tree.ts.  If you can help me mock the mHistory.getSessions part I think I could write tests for these but I don't see it done anywhere in the test code.
image

**Review by lauren-li — DISMISSED** [review]
This looks good to me now. All unit and expected integration tests are passing. With manual testing, I no longer see the issues we were running into before, and the overall experience is much smoother. The views' trees no longer collapse and lose validation status every time a profile is added/hidden or an item is favorited.

One issue I noticed (that is also present in master branch) is if the user validates a profile in the Jobs view by clicking the little arrow next to the profile name, the green circle does not appear even if the profile successfully validates. This not a regression from this PR, and can be addressed in a future issue/PR:
Zowe Explorer - Jobs validation

Thanks @JillieBeanSim for all of your hard work on this!

**Review by crawr — DISMISSED** [review]
LGTM! Tested in Theia and I can confirm the issue I found has been fixed. I also tested the issues found by Lauren and Jelly and they now work as expected. 

Thanks for adding this feature.

**Comment by jellypuno** [comment]
Looks great! Thank you @JillieBeanSim ! I found minor issues but it is not really big deal because the validation is still happening.
I am approving this PR. We can just keep these issues in mind for future re-factoring.

Minor Issue:
- When you Enable Validation and click the PDS the validation is triggered but the folder is not updated.
- When I click Disable Validation and right click again the text is not updated to Enable unless I do something.
- The Greater Than sign next to the profile is not triggering validation. But this is okay because I do want flipstate() to contain the validation because it is called everywhere. It will have more errors.

valid8errz

**Review by jellypuno — DISMISSED** [review]

**Comment by zFernand0** [comment]
This PR gets rid of some performance issues for me, i.e. some random `rg` processes get spawned that consume a lot of CPU do not appear when validation is disabled (on my machine) 😋 
LGTM as well.
Good to go?

**Review by zFernand0 — APPROVED** [review]
Thanks for fixing conflicts
Please ignore https://wash.zowe.org:8443/blue/organizations/jenkins/zowe-cli-vscode-extension/detail/PR-979/27/pipeline/160
Seem like the version of npm ignored completely the `--production` flag and pointed out some dev dependencies vulnerabilities.

**Review by jellypuno — APPROVED** [review]

**Pull request merged**